sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/GlobalSequenceTracker.java

```diff
@@ -0,0 +1,103 @@
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.TimestampedValue;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.joda.time.Duration;
+
+/**
+ * PTransform to produce the side input of the maximum contiguous range of sequence numbers.
+ *
+ * @param <EventKeyT> type of event key
+ * @param <EventT> type of event
+ * @param <ResultT> type of processing result
+ * @param <StateT> type of state
+ */
+class GlobalSequenceTracker<
+        EventKeyT, EventT, ResultT, StateT extends MutableState<EventT, ResultT>>
+    extends PTransform<
+        PCollection<TimestampedValue<KV<EventKeyT, KV<Long, EventT>>>>,
+        PCollectionView<ContiguousSequenceRange>> {
+
+  private final Combine.GloballyAsSingletonView<
+          TimestampedValue<KV<EventKeyT, KV<Long, EventT>>>, ContiguousSequenceRange>
+      sideInputProducer;
+  private final @Nullable Duration frequencyOfGeneration;
+
+  public GlobalSequenceTracker(
+      Combine.GloballyAsSingletonView<
+              TimestampedValue<KV<EventKeyT, KV<Long, EventT>>>, ContiguousSequenceRange>
+          sideInputProducer) {
+    this.sideInputProducer = sideInputProducer;
+    this.frequencyOfGeneration = null;
+  }
+
+  public GlobalSequenceTracker(
+      Combine.GloballyAsSingletonView<
+              TimestampedValue<KV<EventKeyT, KV<Long, EventT>>>, ContiguousSequenceRange>
+          sideInputProducer,
+      Duration globalSequenceGenerationFrequency) {
+    this.sideInputProducer = sideInputProducer;
+    this.frequencyOfGeneration = globalSequenceGenerationFrequency;
+  }
+
+  @Override
+  public PCollectionView<ContiguousSequenceRange> expand(
+      PCollection<TimestampedValue<KV<EventKeyT, KV<Long, EventT>>>> input) {
+    input
+        .getPipeline()
+        .getCoderRegistry()
+        .registerCoderForClass(ContiguousSequenceRange.class, CompletedSequenceRangeCoder.of());
+
+    if (frequencyOfGeneration != null) {
+      // This branch will only be executed in case of streaming pipelines.
+      // For batch pipelines the side input should only be computed once.
+      input =
+          input.apply(
+              "Triggering Setup",
+              // Reproduce the windowing of the input PCollection, but change the triggering
+              // in order to create a slowing changing side input
+              Window.<TimestampedValue<KV<EventKeyT, KV<Long, EventT>>>>into(
+                      (WindowFn<? super TimestampedValue<KV<EventKeyT, KV<Long, EventT>>>, ?>)
+                          input.getWindowingStrategy().getWindowFn())
+                  .accumulatingFiredPanes()
+                  // TODO: verify that we don't need to have the lateness parameterized
+                  .withAllowedLateness(Duration.ZERO)
+                  .triggering(
+                      Repeatedly.forever(
+                          AfterFirst.of(
+                              AfterPane.elementCountAtLeast(1),
```

Comment:
Will this just cause this trigger to fire repeatedly after every element it sees?

Reply:
If so, is there a point in emitting after processing time?

Reply:
I fixed this class - made the number of elements a parameter to the handler (with default of 1000). Unit tests run green.